cpp/src/arrow/array/builder_union.cc

```diff
@@ -19,42 +19,90 @@
-  RETURN_NOT_OK(null_bitmap_builder_.Finish(&null_bitmap));
+  // If the type has not been specified in the constructor, gather type_codes
+  std::vector<uint8_t> type_codes;
+  if (type_ == nullptr) {
+    for (size_t i = 0; i < children_.size(); ++i) {
+      if (type_id_to_children_[i] != nullptr) {
+        type_codes.push_back(static_cast<uint8_t>(i));
+      }
+    }
+  } else {
+    type_codes = checked_cast<const UnionType&>(*type_).type_codes();
+  }
 
-  std::vector<std::shared_ptr<Field>> fields;
-  std::vector<std::shared_ptr<ArrayData>> child_data(children_.size());
-  std::vector<uint8_t> type_ids;
+  std::vector<std::shared_ptr<ArrayData>> child_data(type_codes.size());
   for (size_t i = 0; i < children_.size(); ++i) {
-    std::shared_ptr<ArrayData> data;
-    RETURN_NOT_OK(children_[i]->FinishInternal(&data));
-    child_data[i] = data;
-    fields.push_back(field(field_names_[i], children_[i]->type()));
-    type_ids.push_back(static_cast<uint8_t>(i));
+    RETURN_NOT_OK(children_[i]->FinishInternal(&child_data[i]));
   }
 
   // If the type has not been specified in the constructor, infer it
-  if (!type_) {
-    type_ = union_(fields, type_ids, UnionMode::DENSE);
+  if (type_ == nullptr) {
+    std::vector<std::shared_ptr<Field>> fields;
+    auto field_names_it = field_names_.begin();
+    for (auto&& data : child_data) {
+      fields.push_back(field(*field_names_it++, data->type));
+    }
+    type_ = union_(fields, type_codes, mode_);
   }
 
-  *out = ArrayData::Make(type_, length(), {null_bitmap, types, offsets}, null_count_);
+  *out = ArrayData::Make(type_, length(), {null_bitmap, types, nullptr}, null_count_);
   (*out)->child_data = std::move(child_data);
   return Status::OK();
 }
 
+BasicUnionBuilder::BasicUnionBuilder(
+    MemoryPool* pool, UnionMode::type mode,
+    const std::vector<std::shared_ptr<ArrayBuilder>>& children,
+    const std::shared_ptr<DataType>& type)
+    : ArrayBuilder(type, pool), mode_(mode), types_builder_(pool) {
+  auto union_type = checked_cast<const UnionType*>(type.get());
+  DCHECK_NE(union_type, nullptr);
+  DCHECK_EQ(union_type->mode(), mode);
+
+  children_ = children;
+  type_id_to_children_.resize(union_type->max_type_code() + 1, nullptr);
+
+  auto field_it = type->children().begin();
+  auto children_it = children.begin();
+  for (auto type_id : union_type->type_codes()) {
+    type_id_to_children_[type_id] = *children_it++;
+    field_names_.push_back((*field_it++)->name());
+  }
+  DCHECK_EQ(children_it, children.end());
+  DCHECK_EQ(field_it, type->children().end());
+}
+
+int8_t BasicUnionBuilder::AppendChild(const std::shared_ptr<ArrayBuilder>& new_child,
+                                      const std::string& field_name) {
+  // force type inferrence in Finish
+  type_ = nullptr;
+
+  field_names_.push_back(field_name);
+  children_.push_back(new_child);
+
+  for (; static_cast<size_t>(dense_type_id_) < type_id_to_children_.size();
```

Comment:
I'd DCHECK we reached (U)INT8_MAX . I think it's worthy of comment as it's not exactly clear what and why it's doing.

Reply:
I'll add a comment